### PR TITLE
sql: Add exclude_data_from_backup storage option to ALTER TABLE

### PR DESCRIPTION
**Pull request by adityamaru**
This change adds a storage parameter `exclude_data_from_backup`
to mark a table's row data as excluded from a backup.

`ALTER TABLE foo SET (exclude_data_from_backup = true);`

The above ALTER TABLE stmt sets the `ExcludeDataFromBackup` field
on a table descriptor

ExcludeDataFromBackup specifies if the table's row data can be excluded
from a backup targeting this table. This in turn means that the protected
timestamp record written during the backup will exclude this table as a
target, thereby not holding up GC on this table for the runtime of the
backup. This is particularly useful for table's with high-churn that can
then be configured with a low GC-TTL.

It is important to note that we will still backup the table descriptor, and
all other objects that reference this table egs: jobs, and so restoring
such a table will result in an empty table on the restoring cluster.

This change does not teach any part of the system to use this information
from the table descriptor, that will come in follow up PRs.

Informs: #73536

Release note (sql change): Adds new `ALTER TABLE <table> SET (exclude_data_from_backup = true);`
syntax. This allows a user to exclude row data for a particular table during a backup. On restoring this backup, the table will be created in the restoring cluster but will be empty. This flag also prevents a cluster or database backup from holding up the garbage collection time to live (GC TTL) on the key span representing this table. This is particularly useful for high churn tables that wish to keep a much shorter GC TTL to prevent the buildup of garbage.

### CODE REVIEW

**Comment by adityamaru** [comment]
@otan I'm going to hold off on merging this until your commits have landed so that I don't step on your toes. I just wanted some syntax in place so that I can build https://github.com/cockroachdb/cockroach/pull/75451 on top of this.

**Comment by otan** [comment]
> @otan I'm going to hold off on merging this until your commits have landed so that I don't step on your toes

SGTM!

File: pkg/sql/paramparse/paramobserver.go
Comment (adityamaru) [reply pkg/sql/paramparse/paramobserver.go:206]: oh sweet, removed todo.

File: pkg/sql/logictest/testdata/logic_test/exclude_data_from_backup
Comment (adityamaru) [reply pkg/sql/logictest/testdata/logic_test/exclude_data_from_backup:14]: nice, doing that now.

File: pkg/sql/show_create.go
Comment (otan) [inline pkg/sql/show_create.go:185]: can you grab https://github.com/cockroachdb/cockroach/commit/8d425f6c012072c82a4d3b43c95daa316956170c and make this line `storageParams = append(storageParams, `exclude_data_from_backup = true`)` instead? this will mean we don't double print WITH clauses for TTL + excluding data from backup

File: pkg/sql/paramparse/paramobserver.go
Comment (otan) [inline pkg/sql/paramparse/paramobserver.go:215]: do you want to allow `'on'` as well, as we do with outer variables?

if so, you may want to pull in https://github.com/cockroachdb/cockroach/commit/3bc72cfa9b2c46466f7a33086747ea84592271b3 and use `boolFromDatum`

**Review by otan — APPROVED** [review]
minor suggestions with easy fixes; see #75779 if you need a ref to pull from.

File: pkg/sql/paramparse/paramobserver.go
Comment (adityamaru) [reply pkg/sql/paramparse/paramobserver.go:215]: Hmm not sure  `exclude_data_from_backup = on` makes sense so leaving this as is for now

Comment (otan) [reply pkg/sql/paramparse/paramobserver.go:215]: it also allows `'true'` as a string, and is standard for session vars (e.g. `set experimental_enable_temp_tables = 'on'`)

Comment (otan) [reply pkg/sql/paramparse/paramobserver.go:215]: (although at the end of the day up to you)

Comment (adityamaru) [reply pkg/sql/paramparse/paramobserver.go]: done! sprinkled 'on' and 'true' in the logictest too.

**Comment by adityamaru** [comment]
TFTR!

bors r=otan

**Comment by adityamaru** [comment]
cc: @kathancox I just updated this Release Note now that the feature is complete and being shipped in 22.1.

**Comment by kathancox** [comment]
Hey @adityamaru, if you could open a docs issue that would be great. It would be good to start this in the next week!

**Pull request merged**
